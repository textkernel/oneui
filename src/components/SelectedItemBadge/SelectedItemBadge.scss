@use 'sass:map';

@mixin common-button-style(
    $hoverColor: var(--color-background-neutral-subtlest-hover),
    $activeColor: var(--color-background-neutral-subtlest-pressed)
) {
    & {
        width: var(--space-400);
        height: var(--space-400);
        align-items: center;
        background-color: var(--transparent);
        cursor: pointer;
        display: flex;
        flex-shrink: 0;
        justify-content: center;
    }

    &:hover:not([disabled]) {
        background-color: $hoverColor;
    }
    &:active:not([disabled]) {
        background-color: $activeColor;
    }

    &[disabled] {
        cursor: not-allowed;
    }
}

@mixin selected-state(
    $selectedColor: var(--color-background-selected-subtlest-default),
    $hoverColor: var(--color-background-selected-subtlest-hover),
    $activeColor: var(--color-background-selected-subtlest-pressed)
) {
    & {
        background-color: $selectedColor;
    }
    &:hover:not([disabled]) {
        background-color: $hoverColor;
    }
    &:active:not([disabled]) {
        background-color: $activeColor;
    }
    &[disabled] {
        background-color: $selectedColor;
    }
}

$icon-status-colors: (
    mandatory: (
        normal: var(--color-icon-success-default),
        disabled: var(--color-icon-success-disabled),
    ),
    important: (
        normal: var(--color-icon-cautious-default),
        disabled: var(--color-icon-cautious-disabled),
    ),
    optional: (
        normal: var(--color-icon-subtle),
        disabled: var(--color-icon-disabled),
    ),
    exclude: (
        normal: var(--color-icon-critical-default),
        disabled: var(--color-icon-critical-disabled),
    ),
);

.SelectedItemBadge {
    width: 100%;
    height: var(--space-400);
    border: 1px solid var(--color-border-subtle);
    border-radius: var(--space-100);
    display: flex;
    align-items: center;
    overflow: hidden;

    &__priorityButton {
        padding: 0;
        border: 0;
        border-right: 1px solid var(--color-border-subtle);

        @include common-button-style();
    }

    &__icon {
        height: 20px;
        width: 20px;

        @each $status, $colors in $icon-status-colors {
            &--#{$status} {
                fill: map.get($colors, normal);

                &[disabled] {
                    fill: map.get($colors, disabled);
                }
            }
        }
    }

    &__optionText {
        max-width: var(--space-600);
        min-width: var(--space-200);
        color: var(--color-text-info-default);
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
        flex-shrink: 0;
    }

    &__optionButton {
        width: 100%;
        padding: 0 var(--space-100);
        border: 0;
        flex-grow: 1;
        gap: var(--space-50);

        &[disabled] {
            .SelectedItemBadge__optionText {
                color: var(--color-text-info-disabled);
            }
        }

        @include common-button-style();
    }

    &__priorityButton--isSelected,
    &__optionButton--isSelected {
        @include selected-state();
    }

    &__valueContainer {
        width: 100%;
        padding: 0 var(--space-100);
        float: left;
        display: flex;
        justify-content: center;
        gap: var(--space-50);
        overflow: auto;
        cursor: default;
    }

    &__valueText {
        max-width: 136px;
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
    }

    .badgeDropdownList {
        width: 100%;
        border-radius: var(--space-100);
        overflow: hidden;
    }

    .badgeListItem {
        display: flex;
        align-items: center;
        justify-content: flex-start;
        gap: var(--space-100);
    }

    &__deleteButton {
        border: 0;
        border-left: 1px solid var(--color-border-subtle);
        padding: 0;

        & svg {
            fill: var(--color-icon-subtle);
        }

        &[disabled] {
            & svg {
                fill: var(--color-icon-disabled);
            }
        }

        @include common-button-style();
    }
}
