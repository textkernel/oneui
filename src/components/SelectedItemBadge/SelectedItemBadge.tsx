import * as React from 'react';
import Close from '@material-design-icons/svg/round/close.svg';
import KeyboardDoubleArrowUp from '@material-design-icons/svg/round/keyboard_double_arrow_up.svg';
import KeyboardArrowUp from '@material-design-icons/svg/round/keyboard_arrow_up.svg';
import KeyboardArrowDown from '@material-design-icons/svg/round/keyboard_arrow_down.svg';
import { EmptyElement } from '@textkernel/oneui/customTypes/types';
import { bem } from '../../utils';
import { Dropdown } from '../Dropdown';
import { ListItem, ListItemProps } from '../List';
import { Text } from '../Text';
import styles from './SelectedItemBadge.scss';

const iconMap = {
    mandatory: KeyboardDoubleArrowUp,
    important: KeyboardArrowUp,
    optional: KeyboardArrowDown,
    exclude: Close,
};

export type Priority = 'mandatory' | 'important' | 'optional' | 'exclude';

export type PriorityItem<PriorityItemValue> = {
    priority: Priority;
    label: string;
    value?: PriorityItemValue;
};

// Priority related props
type PriorityProps<PriorityItemValue> = {
    /** Currently selected priority item that indicates the importance of the component. */
    selectedItem: PriorityItem<PriorityItemValue>;
    /** Array of availible priority items. */
    list: Array<PriorityItem<PriorityItemValue>>;
    /** Callback function triggered when a new priority is selected. */
    onChange: (newPriorityItem: PriorityItem<PriorityItemValue>) => void;
    /** Priority button label name for ARIA labelling */
    buttonAriaLabel: string;
};

export interface Props<PriorityItemValue, ChildrenItemValue>
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /**
     * Children nodes to be rendered within the component,
     * specifically used to display the selected value from parent component.
     */
    children?:
        | React.ReactElement<ListItemProps>
        | (React.ReactElement<ListItemProps> | EmptyElement)[]
        | (
              | React.ReactElement<ListItemProps>
              | React.ReactElement<ListItemProps>[]
              | EmptyElement
          )[];
    /** Label of the currently selected option item from filter */
    label: React.ReactNode;
    /** Additional label represents radius or amount of synonyms */
    additionalLabel?: React.ReactNode;
    /**
     * Callback called on selecting one of the passed as children items.
     * Value parameter it is a `value` attribute of children item ({@link ListItemProps.value}).
     */
    onChange?: (value: ChildrenItemValue) => void;
    /** Determines if the children represent a multi-select dropdown */
    isMultiSelect?: boolean;
    /** Function to be called when the delete button is clicked. */
    onDelete?: (e: React.KeyboardEvent | React.MouseEvent) => void;
    /** Boolean indicating whether the whole badge should be disabled. */
    isDisabled?: boolean;
    /** Delete button label name for ARIA labelling */
    deleteButtonLabel?: string;
    priority?: PriorityProps<PriorityItemValue>;
}

const { block, elem } = bem('SelectedItemBadge', styles);
export function SelectedItemBadge<PriorityItemValue, ChildrenItemValue>({
    children,
    label,
    additionalLabel,
    isMultiSelect = false,
    isDisabled = false,
    onChange,
    onDelete = undefined,
    deleteButtonLabel,
    priority,
    ...rest
}: Props<PriorityItemValue, ChildrenItemValue>) {
    const [dropdownStates, setDropdownStates] = React.useState({
        priority: false,
        main: false,
    });

    const badgeRef = React.useRef<HTMLDivElement | null>(null);

    const hasPriorityList = priority && priority.list && priority.list.length > 0;

    const renderPriorityIcon = (priorityType?: Priority, disabled: boolean = false) => {
        if (!priorityType) {
            return null;
        }

        const IconComponent = iconMap[priorityType];

        return IconComponent ? (
            <IconComponent
                {...elem('icon', { [priorityType]: true })}
                disabled={disabled}
                viewBox="0 0 24 24"
            />
        ) : null;
    };

    const handleOnDelete = (e: React.KeyboardEvent | React.MouseEvent) => {
        e.stopPropagation();
        onDelete?.(e);
    };

    const toggleDropdown = (type: 'priority' | 'main', isOpen: boolean) => {
        /**
         * Updates the dropdown state only if 'isOpen' is explicitly defined,
         *  unintended state changes happen by mouse events cause 'isOpen' to become undefined
         * */
        if (isOpen !== undefined) {
            setDropdownStates((prev) => ({ ...prev, [type]: isOpen }));
        }
    };

    return (
        <div {...rest} {...block({ ...rest })} ref={badgeRef}>
            {hasPriorityList && (
                <Dropdown<PriorityItem<PriorityItemValue>>
                    button={
                        <button
                            {...elem('priorityButton', {
                                isSelected: dropdownStates.priority,
                            })}
                            aria-label={priority.buttonAriaLabel}
                            disabled={isDisabled}
                            type="button"
                        >
                            {renderPriorityIcon(priority.selectedItem?.priority, isDisabled)}
                        </button>
                    }
                    additionalSelectProps={{
                        onStateChange: (state) => toggleDropdown('priority', state.isOpen),
                    }}
                    onChange={(newPriorityItem) => priority.onChange(newPriorityItem)}
                    placement="bottom-start"
                    listClassName={styles.badgeDropdownList}
                    refElement={badgeRef}
                >
                    {priority.list.map((item) => (
                        <ListItem className={styles.badgeListItem} key={item.priority} value={item}>
                            {renderPriorityIcon(item.priority)}
                            <Text inline size="small">
                                {item.label}
                            </Text>
                        </ListItem>
                    ))}
                </Dropdown>
            )}

            {children ? (
                <Dropdown<ChildrenItemValue>
                    button={
                        <button
                            {...elem('optionButton', {
                                isSelected: dropdownStates.main,
                            })}
                            {...(typeof label === 'string' && { 'aria-label': label })}
                            disabled={isDisabled}
                            type="button"
                        >
                            <Text
                                inline
                                size="small"
                                {...elem('valueText')}
                                {...(typeof label === 'string' && { title: label })}
                            >
                                {label}
                            </Text>
                            {additionalLabel && (
                                <Text
                                    {...elem('optionText')}
                                    inline
                                    {...(typeof additionalLabel === 'string' && {
                                        title: additionalLabel,
                                    })}
                                    size="small"
                                >
                                    {additionalLabel}
                                </Text>
                            )}
                        </button>
                    }
                    additionalSelectProps={{
                        onStateChange: (state) => toggleDropdown('main', state.isOpen),
                    }}
                    onChange={(newOptionItem) => onChange?.(newOptionItem)}
                    placement="bottom"
                    refElement={badgeRef}
                    listClassName={styles.badgeDropdownList}
                    isMultiSelect={isMultiSelect}
                >
                    {children}
                </Dropdown>
            ) : (
                <div {...elem('valueContainer')}>
                    <Text
                        inline
                        size="small"
                        {...elem('valueText')}
                        {...(typeof label === 'string' && { title: label })}
                    >
                        {label}
                    </Text>
                    {additionalLabel && (
                        <Text
                            {...elem('optionText')}
                            inline
                            {...(typeof additionalLabel === 'string' && {
                                title: additionalLabel,
                            })}
                            size="small"
                        >
                            {additionalLabel}
                        </Text>
                    )}
                </div>
            )}

            {onDelete && (
                <button
                    {...elem('deleteButton')}
                    aria-label={deleteButtonLabel}
                    disabled={isDisabled}
                    onClick={handleOnDelete}
                    type="button"
                >
                    <Close
                        viewBox="0 0 24 24"
                        style={{
                            width: '20px',
                            height: '20px',
                        }}
                    />
                </button>
            )}
        </div>
    );
}

SelectedItemBadge.displayName = 'SelectedItemBadge';
