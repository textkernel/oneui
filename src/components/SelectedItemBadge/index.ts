export { SelectedItemBadge, Props as SelectedItemBadgeProps } from './SelectedItemBadge';
