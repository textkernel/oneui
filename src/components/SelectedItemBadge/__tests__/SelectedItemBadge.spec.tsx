import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import { SelectedItemBadge } from '../SelectedItemBadge';
import { SelectedItemBadgeProps } from '..';

const mockOnPriorityItemChange = jest.fn();
const mockOnDelete = jest.fn();

const defaultProps: SelectedItemBadgeProps<string, string> = {
    label: 'London',
    priority: {
        selectedItem: { priority: 'mandatory', label: 'Mandatory', value: 'required' },
        list: [
            { priority: 'mandatory', label: 'Mandatory', value: 'required' },
            { priority: 'important', label: 'Important', value: 'strongly_favored' },
            { priority: 'optional', label: 'Optional', value: 'favored' },
            { priority: 'exclude', label: 'Exclude', value: 'banned' },
        ],
        buttonAriaLabel: 'priority button',
        onChange: mockOnPriorityItemChange,
    },
    deleteButtonLabel: 'delete button',
    onDelete: mockOnDelete,
};

const renderSelectedItemBadge = (props = {}) =>
    render(<SelectedItemBadge {...defaultProps} {...props} />);

describe('SelectedItemBadge', () => {
    describe('Priority button', () => {
        it('renders correctly with all props provided', () => {
            const view = renderSelectedItemBadge();
            expect(view.container).toMatchSnapshot();
        });

        it('should handle priority change correctly', async () => {
            const view = renderSelectedItemBadge();
            const user = userEvent.setup();
            const priorityButton = screen.getByLabelText('priority button');
            await user.click(priorityButton);
            expect(view.container).toMatchSnapshot();
            await user.click(screen.getByText('Important'));
            expect(mockOnPriorityItemChange).toHaveBeenCalledTimes(1);
            expect(mockOnPriorityItemChange).toHaveBeenCalledWith({
                priority: 'important',
                label: 'Important',
                value: 'strongly_favored',
            });
        });

        it('does not render the priority button when priority is not provided', () => {
            const { queryByLabelText } = renderSelectedItemBadge({
                priority: undefined,
            });
            expect(queryByLabelText('priority button')).not.toBeInTheDocument();
        });

        it('does not render the priority button when priority list is not provided', () => {
            const { queryByLabelText } = renderSelectedItemBadge({
                priority: { list: undefined },
            });
            expect(queryByLabelText('priority button')).not.toBeInTheDocument();
        });

        it('does not render the priority button when priority list is empty', () => {
            const { queryByLabelText } = renderSelectedItemBadge({
                priority: { list: [] },
            });
            expect(queryByLabelText('priority button')).not.toBeInTheDocument();
        });
    });

    describe('Option button', () => {
        it.skip('renders children list headline when provided', async () => {});

        it.skip('should handle change correctly', async () => {});

        it.skip('does not render the option button but displays plain text with passed children if option is not provided', () => {});

        it.skip('does not render the option button but displays plain text with passed children if option list is not provided', () => {});

        it.skip('does not render the option button but displays plain text with passed children if option list is empty', () => {});
    });

    describe('Delete button', () => {
        it('should not render the delete button if onDelete is not provided', () => {
            const { queryByLabelText } = renderSelectedItemBadge({ onDelete: undefined });
            expect(queryByLabelText('delete button')).not.toBeInTheDocument();
        });

        it('should handle the delete button functionality correctly', async () => {
            const { getByLabelText } = renderSelectedItemBadge();
            const user = userEvent.setup();
            const deleteButton = getByLabelText('delete button');
            await user.click(deleteButton);
            expect(mockOnDelete).toHaveBeenCalled();
        });
    });

    it('renders all the buttons disabled when provided from props', () => {
        const { getByLabelText } = renderSelectedItemBadge({
            isDisabled: true,
        });
        expect(getByLabelText('priority button')).toBeDisabled();
        expect(getByLabelText('delete button')).toBeDisabled();
    });

    it('maintains state correctly after multiple interactions', async () => {
        const { getByText } = renderSelectedItemBadge();
        const user = userEvent.setup();
        const priorityButton = screen.getByLabelText('priority button');
        await user.click(priorityButton); // Open
        await user.click(priorityButton); // Close
        await user.click(priorityButton); // Re-open
        defaultProps?.priority?.list?.forEach(({ label }) => {
            expect(screen.getByText(label)).toBeInTheDocument();
        });
        await user.click(getByText('Optional'));
        expect(defaultProps?.priority?.onChange).toHaveBeenCalledTimes(1);
        expect(defaultProps?.priority?.onChange).toHaveBeenCalledWith({
            label: 'Optional',
            priority: 'optional',
            value: 'favored',
        });
    });
});
