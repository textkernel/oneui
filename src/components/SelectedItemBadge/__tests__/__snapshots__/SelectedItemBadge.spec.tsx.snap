// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedItemBadge Priority button renders correctly with all props provided 1`] = `
<div>
  <div
    class="SelectedItemBadge"
  >
    <button
      aria-activedescendant=""
      aria-controls="downshift-:r0:-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-label="priority button"
      aria-labelledby="downshift-:r0:-label"
      class="SelectedItemBadge__priorityButton"
      id="downshift-:r0:-toggle-button"
      role="combobox"
      tabindex="0"
      type="button"
    >
      <svg
        class="SelectedItemBadge__icon SelectedItemBadge__icon--mandatory"
        data-testid="default-icon"
        viewBox="0 0 24 24"
      />
    </button>
    <ul
      aria-labelledby="downshift-:r0:-label"
      class="List"
      id="downshift-:r0:-menu"
      role="listbox"
    />
    <div
      class="SelectedItemBadge__valueContainer"
    >
      <span
        class="Text--size_small SelectedItemBadge__valueText"
        title="London"
      >
        London
      </span>
    </div>
    <button
      aria-label="delete button"
      class="SelectedItemBadge__deleteButton"
      type="button"
    >
      <svg
        data-testid="default-icon"
        style="width: 20px; height: 20px;"
        viewBox="0 0 24 24"
      />
    </button>
  </div>
</div>
`;

exports[`SelectedItemBadge Priority button should handle priority change correctly 1`] = `
<div>
  <div
    class="SelectedItemBadge"
  >
    <button
      aria-activedescendant=""
      aria-controls="downshift-:r1:-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-label="priority button"
      aria-labelledby="downshift-:r1:-label"
      class="SelectedItemBadge__priorityButton SelectedItemBadge__priorityButton--isSelected"
      id="downshift-:r1:-toggle-button"
      role="combobox"
      tabindex="0"
      type="button"
    >
      <svg
        class="SelectedItemBadge__icon SelectedItemBadge__icon--mandatory"
        data-testid="default-icon"
        viewBox="0 0 24 24"
      />
    </button>
    <ul
      aria-labelledby="downshift-:r1:-label"
      class="List Dropdown__list badgeDropdownList"
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
      id="downshift-:r1:-menu"
      role="listbox"
      style="position: absolute; left: 0px; top: 0px; width: 0px; transform: translate(0px, 3px);"
    >
      <li
        aria-disabled="false"
        aria-selected="false"
        class="ListItem ListItem--clickable List__item badgeListItem"
        id="downshift-:r1:-item-0"
        role="option"
      >
        <svg
          class="SelectedItemBadge__icon SelectedItemBadge__icon--mandatory"
          data-testid="default-icon"
          viewBox="0 0 24 24"
        />
        <span
          class="Text--size_small"
        >
          Mandatory
        </span>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="ListItem ListItem--clickable List__item badgeListItem"
        id="downshift-:r1:-item-1"
        role="option"
      >
        <svg
          class="SelectedItemBadge__icon SelectedItemBadge__icon--important"
          data-testid="default-icon"
          viewBox="0 0 24 24"
        />
        <span
          class="Text--size_small"
        >
          Important
        </span>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="ListItem ListItem--clickable List__item badgeListItem"
        id="downshift-:r1:-item-2"
        role="option"
      >
        <svg
          class="SelectedItemBadge__icon SelectedItemBadge__icon--optional"
          data-testid="default-icon"
          viewBox="0 0 24 24"
        />
        <span
          class="Text--size_small"
        >
          Optional
        </span>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="ListItem ListItem--clickable List__item badgeListItem"
        id="downshift-:r1:-item-3"
        role="option"
      >
        <svg
          class="SelectedItemBadge__icon SelectedItemBadge__icon--exclude"
          data-testid="default-icon"
          viewBox="0 0 24 24"
        />
        <span
          class="Text--size_small"
        >
          Exclude
        </span>
      </li>
    </ul>
    <div
      class="SelectedItemBadge__valueContainer"
    >
      <span
        class="Text--size_small SelectedItemBadge__valueText"
        title="London"
      >
        London
      </span>
    </div>
    <button
      aria-label="delete button"
      class="SelectedItemBadge__deleteButton"
      type="button"
    >
      <svg
        data-testid="default-icon"
        style="width: 20px; height: 20px;"
        viewBox="0 0 24 24"
      />
    </button>
  </div>
</div>
`;
